blocks_verification.md

```diff
@@ -0,0 +1,137 @@
+Blocks are stored on Tanker's infrastructure and distributed to the different clients. 
+
+The tree structure of the Trustchain can be read to reconstruct the state of each user's devices at any given time, and verified to make sure every action taken was legitimate at the time. Such a verification is done by every Tanker client receiving a new block, before extracting any information from their payload. Each block’s signature is verified, then their author’s block’s signature recursively until the root block is reached.
+
+## Definitions
+
+- *a zero-array* is an array filled with zeros
+- *the user has a user key*: this means that the user has at least one block of type DeviceCreation 3 or DeviceRevocation2
+
+## Blocks
+
+- The author field must be a TrustchainCreation or a DeviceCreation (except for TrustchainCreation)
+- The signature field must contain the signature of the hash of the block by the private key of the author (except for TrustchainCreation and DeviceCreation)
+
+## Trustchain Creation
+
+Verification:
+
+- The author of the block must be a zero-array
+- The signature of the block must be a zero-array
+- There must be no block before this one
+- The hash of the block must match the Trustchain ID
+
+Server side:
+
+- The Trustchain signature key must be unique
+
+## Device Creations
+
+Verification:
+
+- The user_id must be the same as the author's one (in case of a secondary device)
+- The delegation_signature field must contain the signature of the delegation (user_id + ephemeral public key) by the author's key
+- The block's signature field must contain the signature of the block by the ephemeral key
+
+Server-side verifications:
+
+- The user_id must not already exist on the Trustchain
+- The device public signature and encryption key must be unique
+
+### DeviceCreation v1
+
+### DeviceCreation v2
+
+- The user must not have a user key
+
+### DeviceCreation v3
+
+- The user PublicEncryptionKey must be unique
+- It cannot change the user PublicEncryptionKey (if the user already has one)
+
+## Device Revocations
+
+Verification:
+
+- The author must be a DeviceCreation
+- The device_id field must contain the hash of a DeviceCreation block
+- The revoked device must belong to the author
+- The revoked device must not be already revoked
+
+A device can revoke itself, even if this results in all a user's devices being revoked.
```

Comment:
Did you mean that, if it's the last of its device, he's locked himself out ? Somehow that come as is one device revocation would revoke all of them ....